core/src/main/java/org/elasticsearch/cluster/metadata/IndexCreationMonitor.java

```diff
@@ -0,0 +1,131 @@
+ * desired for the index creation operation.
+ */
+public final class IndexCreationMonitor extends AbstractComponent {
+
+    private final ClusterService clusterService;
+    private final ThreadContext threadContext;
+
+    protected IndexCreationMonitor(Settings settings, ClusterService clusterService, ThreadContext threadContext) {
+        super(settings);
+        this.clusterService = Objects.requireNonNull(clusterService);
+        this.threadContext = Objects.requireNonNull(threadContext);
+    }
+
+    void waitOnShards(final String indexName,
+                      final int waitOnActiveShards,
+                      final ActionListener<ClusterStateUpdateResponse> listener,
+                      final TimeValue timeout) {
+
+        final ClusterStateObserver observer = new ClusterStateObserver(clusterService, logger, threadContext);
+        final ClusterState observedState = observer.observedState();
+        final ClusterStateObserver.ChangePredicate shardsAllocatedPredicate = new ClusterStateObserver.ChangePredicate() {
+            @Override
+            public boolean apply(ClusterState previousState, ClusterState.ClusterStateStatus previousStatus,
+                                 ClusterState newState, ClusterState.ClusterStateStatus newStatus) {
+                return newStatus == ClusterState.ClusterStateStatus.APPLIED
+                           && enoughShardsActive(newState, indexName, settings, waitOnActiveShards);
+            }
+
+            @Override
+            public boolean apply(ClusterChangedEvent changedEvent) {
+                return enoughShardsActive(changedEvent.state(), indexName, settings, waitOnActiveShards);
+            }
+        };
+        final ClusterStateObserver.Listener observerListener = new ClusterStateObserver.Listener() {
+            @Override
+            public void onNewClusterState(ClusterState state) {
+                listener.onResponse(new ClusterStateUpdateResponse(true));
+            }
+
+            @Override
+            public void onClusterServiceClose() {
+                logger.warn("[{}] cluster service closed while waiting for write consistency number of shards " +
+                                "to be allocated on index creation.", indexName);
+                listener.onResponse(new ClusterStateUpdateResponse(false));
+            }
+
+            @Override
+            public void onTimeout(TimeValue timeout) {
+                logger.warn("[{}] index was created, but the operation timed out waiting for write consistency " +
+                                "number of shards to be allocated.", indexName);
+                listener.onResponse(new ClusterStateUpdateResponse(false));
+            }
+        };
+
+        if (observedState.status() == ClusterState.ClusterStateStatus.APPLIED
+                && enoughShardsActive(observedState, indexName, settings, waitOnActiveShards)) {
+            observerListener.onNewClusterState(observedState);
```

Comment:
I wonder if repeating the Change predicate here is needed. When we call observer.waitForNextChange() and the predicate holds, we directly call `onNewClusterState` on the listener...